File: solutions/ruby/protein-translation/1/protein_translation.rb
class InvalidCodonError < StandardError; end 

module Translation 
  CODON_AMINO_ACID = {
    "AUG" => "Methionine", "UUU" => "Phenylalanine", "UUC" => "Phenylalanine",
    "UUA" => "Leucine", "UUG" => "Leucine", "UCU" => "Serine", "UCC" => "Serine",
    "UCA" => "Serine", "UCG" => "Serine", "UAU" => "Tyrosine", "UAC" => "Tyrosine",
    "UGU" => "Cysteine", "UGC" => "Cysteine", "UGG" => "Tryptophan", "UAA" => "STOP",
    "UAG" => "STOP", "UGA" => "STOP"  
  }.freeze 
  
  def self.of_rna(strand) 
    codons = strand.chars.each_slice(3).map(&:join) 
    stop = codons.index("UAA") || codons.index("UAG") || codons.index("UGA")
    codons = codons[0...stop]
    debug "#{codons}"

    codons.each do |codon|
      raise InvalidCodonError unless CODON_AMINO_ACID.key?(codon)
    end 
    
    codons.map do |codon| 
      CODON_AMINO_ACID[codon] 
    end
  end 
end 

File: solutions/cpp/protein-translation/1/protein_translation.cpp
#include "protein_translation.h"

namespace protein_translation {
    vector<string> proteins(string rna) {
        vector<string> p;
        
        for (size_t i = 0; i < rna.size(); i+=3) {
            string codon = rna.substr(i, 3);
            
            if (codon == "AUG") {
                p.push_back("Methionine");
            } else if (codon == "UUU" || codon == "UUC") {
                p.push_back("Phenylalanine");
            } else if(codon == "UUA" || codon == "UUG") {
                p.push_back("Leucine");
            } else if(codon == "UCU" || codon == "UCC" || codon == "UCA" || codon == "UCG") {
                p.push_back("Serine");
            } else if (codon == "UAU" || codon == "UAC") {
                p.push_back("Tyrosine");
            } else if (codon == "UGU" || codon == "UGC") {
                p.push_back("Cysteine");
            } else if (codon == "UGG") {
                p.push_back("Tryptophan");
            } else if (codon == "UAA" || codon == "UAG" || codon == "UGA") {
                break; 
            }
        }

        return p;
    }
}  // namespace protein_translation
